ci/ray_ci/clean_up_old_docker_commit_tags.py

```diff
@@ -0,0 +1,193 @@
+    # The URL for token authentication
+    url = f"https://auth.docker.io/token?service={service}&scope={scope}"
+    response = requests.get(url)
+    token = response.json().get("token")
+    return token
+
+
+def count_docker_tags():
+    """
+    Count number of tags from rayproject/ray repository.
+    """
+    response = requests.get(
+        "https://hub.docker.com/v2/namespaces/rayproject/repositories/ray/tags"
+    )
+    tag_count = response.json()["count"]
+    return tag_count
+
+
+def get_image_creation_time(repository: str, tag: str):
+    """
+    Get the creation time of the image from the tag image config.
+    """
+    res = subprocess.run(
+        ["crane", "config", f"{repository}:{tag}"], capture_output=True, text=True
+    )
+    if res.returncode != 0 or not res.stdout:
+        return None
+    manifest = json.loads(res.stdout)
+    created = manifest["created"]
+    created_time = datetime.fromisoformat(created)
+    return created_time
+
+
+def get_auth_token_docker_hub(username: str, password: str):
+    params = {
+        "username": username,
+        "password": password,
+    }
+    headers = {
+        "Content-Type": "application/json",
+    }
+    response = requests.post(
+        "https://hub.docker.com/v2/users/login", headers=headers, params=params
+    )
+    token = response.json().get("token")
+    return token
+
+
+def delete_tags(namespace: str, repository: str, tags: list[str]):
+    """
+    Delete tag from Docker Hub repo.
+    """
+    token = get_auth_token_docker_hub("username", "password")
+    headers = {
+        "Authorization": f"Bearer {token}",
+    }
+    for tag in tags:
+        print(f"Deleting {tag}")  # TODO: delete this line
+        url = f"https://hub.docker.com/v2/repositories/{namespace}/{repository}/tags/{tag}"
+        response = requests.delete(url, headers=headers)
+        if response.status_code != 204:
+            print(f"Failed to delete {tag}")
+
+
+def query_tags_to_delete(
+    page_count: int, commit_short_shas: Set[str], page_size: int = 100
+):
+    """
+    Query tags to delete from rayproject/ray repository.
+    """
+    get_docker_token()
+    repository = "rayproject/ray"
+    current_time = datetime.now(timezone.utc)
+    tags_to_delete = []
+    for page in range(page_count, 0, -1):
+        print("Querying page ", page)  # Replace with log
+        print("Delete count: ", len(tags_to_delete))  # Replace with log
+        response = requests.get(
+            "https://hub.docker.com/v2/namespaces/rayproject/repositories/ray/tags",
+            params={"page": page, "page_size": 100}
+        )
+        if "errors" in response.json():
+            continue
+        result = response.json()["results"]
+        tags = [tag["name"] for tag in result]
+        # Check if tag is in list of commit SHAs
+        commit_tags = [
+            tag
+            for tag in tags
+            if len(tag.split("-")[0]) == 6 and tag.split("-")[0] in commit_short_shas
```

Comment:
you might want to be careful where there might be duplicates in `commit_short_shas`, that is, a recent commit (within 30 days) might share a same commit prefix with one from the past.

our guarding / criteria for delete or not should be something like this:

````
if it is recent (created / uploaded) within N days:
  do not delete
else if it has a commit prefix:
  if the commit prefix is within N days of recent commits:
    do not delete
  else:
    delele
````